### ISSUE TITLE
Docker volumes issue

### ISSUE THREAD

**Issue by czerasz** [ansible/ansible-modules-core#4228]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

**docker_container** module
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = 
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

No custom configuration.
##### OS / ENVIRONMENT

Using Ansible inside a container from [`williamyeh/ansible:debian8`](https://hub.docker.com/r/williamyeh/ansible/)
##### SUMMARY

I want to start a container with DataDog agent inside. The agents needs some host files which I mount via volumes.
##### STEPS TO REPRODUCE

Here is the playbook I try to provision:

```
- hosts: all
  tasks:
  - name: Start DataDog agent inside container
    docker_container:
      api_version: 1.18
      name: dd-agent
      hostname: "{{ facter_hostname }}"
      image: "datadog/docker-dd-agent:latest"
      state: started
      restart_policy: always
      restart: yes
      env:
        API_KEY: "{{ datadog_api_key }}"
      volumes:
        - /var/run/docker.sock:/var/run/docker.sock
        - /proc:/host/proc:ro
        - /sys/fs/cgroup:/host/sys/fs/cgroup:ro
        - "{{ ansible_env.HOME }}/dd-agent:/conf.d:ro"
```
##### EXPECTED RESULTS

Docker container was started
##### ACTUAL RESULTS

I get the following issue:

```
fatal: [hostname]: FAILED! => 

{
  "changed": false,
  "failed": true,
  "invocation": {
    "module_args": {
      "api_version": "1.18",
      "blkio_weight": null,
      "cacert_path": null,
      "capabilities": null,
      "cert_path": null,
      "command": null,
      "cpu_period": null,
      "cpu_quota": null,
      "cpu_shares": null,
      "cpuset_cpus": null,
      "cpuset_mems": null,
      "debug": false,
      "detach": true,
      "devices": null,
      "dns_opts": null,
      "dns_search_domains": null,
      "dns_servers": null,
      "docker_host": null,
      "entrypoint": null,
      "env": {
        "API_KEY": "<secret>"
      },
      "etc_hosts": null,
      "exposed_ports": null,
      "filter_logger": false,
      "force_kill": false,
      "groups": null,
      "hostname": "hostname",
      "image": "datadog/docker-dd-agent:latest",
      "interactive": false,
      "ipc_mode": null,
      "keep_volumes": true,
      "kernel_memory": null,
      "key_path": null,
      "kill_signal": null,
      "labels": null,
      "links": null,
      "log_driver": "json-file",
      "log_options": null,
      "mac_address": null,
      "memory": "0",
      "memory_reservation": null,
      "memory_swap": null,
      "memory_swappiness": null,
      "name": "dd-agent",
      "network_mode": null,
      "networks": null,
      "oom_killer": null,
      "paused": false,
      "pid_mode": null,
      "privileged": false,
      "published_ports": null,
      "pull": false,
      "read_only": false,
      "recreate": false,
      "restart": true,
      "restart_policy": "always",
      "restart_retries": 0,
      "security_opts": null,
      "shm_size": null,
      "ssl_version": null,
      "state": "started",
      "stop_signal": null,
      "stop_timeout": null,
      "timeout": null,
      "tls": null,
      "tls_hostname": null,
      "tls_verify": null,
      "trust_image_content": false,
      "tty": false,
      "ulimits": null,
      "user": null,
      "uts": null,
      "volume_driver": null,
      "volumes": [
        "/var/run/docker.sock:/var/run/docker.sock",
        "/proc:/host/proc:ro",
        "/sys/fs/cgroup:/host/sys/fs/cgroup:ro",
        "/home/user/dd-agent:/conf.d:ro"
      ],
      "volumes_from": null
    },
    "module_name": "docker_container"
  },
  "msg": "Error starting container 16841e038b6fac0786e0c755d778a18d0602ab820362b511d0a4dfb72fada2ee: 500 Server Error: Internal Server Error (\"Cannot start container 16841e038b6fac0786e0c755d778a18d0602ab820362b511d0a4dfb72fada2ee: [8] System error: \"/var/lib/docker/aufs/mnt/16841e038b6fac0786e0c755d778a18d0602ab820362b511d0a4dfb72fada2ee/proc\" cannot be mounted because it is located inside \"/proc\"\")"
}
```

I'm able to start the container using the `shell` module:

```
- hosts: all
  tasks:
  - name: Start DataDog agent inside container
    shell: |
      docker run -d --name dd-agent \
                 -h {{ facter_hostname }} \
                 -v /var/run/docker.sock:/var/run/docker.sock \
                 -v /proc/:/host/proc/:ro \
                 -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
                 -v {{ ansible_env.HOME }}/dd-agent:/conf.d:ro \
                 -e API_KEY={{ datadog_api_key }} \
                 --restart=always \
                 datadog/docker-dd-agent:latest
```

Here is the output of `docker inspect dd-agent` for the container started with the `docker_container` module:

```
[{
    "AppArmorProfile": "",
    "Args": [
        "supervisord",
        "-n",
        "-c",
        "/etc/dd-agent/supervisor.conf"
    ],
    "Config": {
        "AttachStderr": false,
        "AttachStdin": false,
        "AttachStdout": false,
        "Cmd": [
            "supervisord",
            "-n",
            "-c",
            "/etc/dd-agent/supervisor.conf"
        ],
        "CpuShares": 0,
        "Cpuset": "",
        "Domainname": "",
        "Entrypoint": [
            "/entrypoint.sh"
        ],
        "Env": [
            "API_KEY=<secret>",
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "DOCKER_DD_AGENT=yes",
            "AGENT_VERSION=1:5.8.4-1"
        ],
        "ExposedPorts": {
            "8125/udp": {},
            "9001/tcp": {}
        },
        "Hostname": "hostname",
        "Image": "datadog/docker-dd-agent:latest",
        "Labels": {},
        "MacAddress": "",
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkDisabled": false,
        "OnBuild": null,
        "OpenStdin": false,
        "PortSpecs": null,
        "StdinOnce": false,
        "Tty": false,
        "User": "",
        "Volumes": {
            "/checks.d": {},
            "/conf.d": {},
            "/home/user/dd-agent": {},
            "/proc": {},
            "/sys/fs/cgroup": {},
            "/var/run/docker.sock": {}
        },
        "WorkingDir": ""
    },
    "Created": "2016-07-21T11:58:11.339087473Z",
    "Driver": "aufs",
    "ExecDriver": "native-0.2",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": [
            "/var/run/docker.sock:/var/run/docker.sock:rw",
            "/home/user/dd-agent:/conf.d:ro",
            "/sys/fs/cgroup:/host/sys/fs/cgroup:ro",
            "/proc:/host/proc:ro"
        ],
        "CapAdd": null,
        "CapDrop": null,
        "CgroupParent": "",
        "ContainerIDFile": "",
        "CpuShares": 0,
        "CpusetCpus": "",
        "Devices": null,
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "IpcMode": "",
        "Links": null,
        "LogConfig": {
            "Config": {},
            "Type": "json-file"
        },
        "LxcConf": null,
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkMode": "",
        "PidMode": "",
        "PortBindings": null,
        "Privileged": false,
        "PublishAllPorts": false,
        "ReadonlyRootfs": false,
        "RestartPolicy": {
            "MaximumRetryCount": 0,
            "Name": ""
        },
        "SecurityOpt": null,
        "Ulimits": null,
        "VolumesFrom": null
    },
    "HostnamePath": "/var/lib/docker/containers/46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72/hostname",
    "HostsPath": "/var/lib/docker/containers/46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72/hosts",
    "Id": "46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72",
    "Image": "f75d3383b801f85d22212a836b31623f6230d11c466c1564a137bda2fa9c12cd",
    "LogPath": "/var/lib/docker/containers/46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72/46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72-json.log",
    "MountLabel": "",
    "Name": "/dd-agent",
    "NetworkSettings": {
        "Bridge": "",
        "Gateway": "",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "",
        "IPPrefixLen": 0,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "MacAddress": "",
        "PortMapping": null,
        "Ports": null
    },
    "Path": "/entrypoint.sh",
    "ProcessLabel": "",
    "ResolvConfPath": "/var/lib/docker/containers/46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72/resolv.conf",
    "RestartCount": 0,
    "State": {
        "Dead": false,
        "Error": "[8] System error: \"/var/lib/docker/aufs/mnt/46e60eb37514ec1813eec1eef78f08e2dde967a816dd86ee9f32700975738e72/proc\" cannot be mounted because it is located inside \"/proc\"",
        "ExitCode": -1,
        "FinishedAt": "0001-01-01T00:00:00Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 0,
        "Restarting": false,
        "Running": false,
        "StartedAt": "0001-01-01T00:00:00Z"
    },
    "Volumes": {
        "/checks.d": "/var/lib/docker/vfs/dir/fb8da45e0252592a5655dfe327edbb66fa12acf1226a21611f814ac925381f54",
        "/conf.d": "/home/user/dd-agent",
        "/home/user/dd-agent": "/var/lib/docker/vfs/dir/30560b34ecc6821b23613316cdc88ef0d46545f8492d7fa2b62157f6fb35e46a",
        "/host/proc": "/proc",
        "/host/sys/fs/cgroup": "/sys/fs/cgroup",
        "/proc": "/var/lib/docker/vfs/dir/7a1ba871ced8edbe9b62659efa8980217d68517f7b74cb9751323eb0e0a8a289",
        "/sys/fs/cgroup": "/var/lib/docker/vfs/dir/70015740c833c8c699a963c9b9735bd02c10393afed7dd9e49945f4489f4d5e3",
        "/var/run/docker.sock": "/run/docker.sock"
    },
    "VolumesRW": {
        "/checks.d": true,
        "/conf.d": false,
        "/home/user/dd-agent": true,
        "/host/proc": false,
        "/host/sys/fs/cgroup": false,
        "/proc": true,
        "/sys/fs/cgroup": true,
        "/var/run/docker.sock": true
    }
}
]
```

Here is the output of `docker inspect dd-agent` for the container started with the `shell` module:

```
[{
    "AppArmorProfile": "",
    "Args": [
        "supervisord",
        "-n",
        "-c",
        "/etc/dd-agent/supervisor.conf"
    ],
    "Config": {
        "AttachStderr": false,
        "AttachStdin": false,
        "AttachStdout": false,
        "Cmd": [
            "supervisord",
            "-n",
            "-c",
            "/etc/dd-agent/supervisor.conf"
        ],
        "CpuShares": 0,
        "Cpuset": "",
        "Domainname": "",
        "Entrypoint": [
            "/entrypoint.sh"
        ],
        "Env": [
            "API_KEY=<secret>",
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "DOCKER_DD_AGENT=yes",
            "AGENT_VERSION=1:5.8.4-1"
        ],
        "ExposedPorts": {
            "8125/udp": {},
            "9001/tcp": {}
        },
        "Hostname": "hostname",
        "Image": "datadog/docker-dd-agent:latest",
        "Labels": {},
        "MacAddress": "",
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkDisabled": false,
        "OnBuild": null,
        "OpenStdin": false,
        "PortSpecs": null,
        "StdinOnce": false,
        "Tty": false,
        "User": "",
        "Volumes": {
            "/checks.d": {},
            "/conf.d": {}
        },
        "WorkingDir": ""
    },
    "Created": "2016-07-21T12:55:46.815388898Z",
    "Driver": "aufs",
    "ExecDriver": "native-0.2",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": [
            "/var/run/docker.sock:/var/run/docker.sock",
            "/proc/:/host/proc/:ro",
            "/sys/fs/cgroup/:/host/sys/fs/cgroup:ro",
            "/home/user/dd-agent:/conf.d:ro"
        ],
        "CapAdd": null,
        "CapDrop": null,
        "CgroupParent": "",
        "ContainerIDFile": "",
        "CpuShares": 0,
        "CpusetCpus": "",
        "Devices": [],
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "IpcMode": "",
        "Links": null,
        "LogConfig": {
            "Config": null,
            "Type": "json-file"
        },
        "LxcConf": [],
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkMode": "bridge",
        "PidMode": "",
        "PortBindings": {},
        "Privileged": false,
        "PublishAllPorts": false,
        "ReadonlyRootfs": false,
        "RestartPolicy": {
            "MaximumRetryCount": 0,
            "Name": "always"
        },
        "SecurityOpt": null,
        "Ulimits": null,
        "VolumesFrom": null
    },
    "HostnamePath": "/var/lib/docker/containers/bb765b010df36a5e4b76f97a01be91f5f922a602958bda532b6b5c3ce0962b71/hostname",
    "HostsPath": "/var/lib/docker/containers/bb765b010df36a5e4b76f97a01be91f5f922a602958bda532b6b5c3ce0962b71/hosts",
    "Id": "bb765b010df36a5e4b76f97a01be91f5f922a602958bda532b6b5c3ce0962b71",
    "Image": "f75d3383b801f85d22212a836b31623f6230d11c466c1564a137bda2fa9c12cd",
    "LogPath": "/var/lib/docker/containers/bb765b010df36a5e4b76f97a01be91f5f922a602958bda532b6b5c3ce0962b71/bb765b010df36a5e4b76f97a01be91f5f922a602958bda532b6b5c3ce0962b71-json.log",
    "MountLabel": "",
    "Name": "/dd-agent",
    "NetworkSettings": {
        "Bridge": "docker0",
        "Gateway": "172.17.42.1",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "172.17.0.xx",
        "IPPrefixLen": 16,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "xxx::42:acff:xxx:xx",
        "LinkLocalIPv6PrefixLen": 64,
        "MacAddress": "02:42:ac:11:xx:xx",
        "PortMapping": null,
        "Ports": {
            "8125/udp": null,
            "9001/tcp": null
        }
    },
    "Path": "/entrypoint.sh",
    "ProcessLabel": "",
    "ResolvConfPath": "/var/lib/docker/containers/bb765b010df36a5e4b76f97a01be91f5f922a602958bda532b6b5c3ce0962b71/resolv.conf",
    "RestartCount": 0,
    "State": {
        "Dead": false,
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "0001-01-01T00:00:00Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 2828,
        "Restarting": false,
        "Running": true,
        "StartedAt": "2016-07-21T12:55:46.932427737Z"
    },
    "Volumes": {
        "/checks.d": "/var/lib/docker/vfs/dir/b6c7c881ee0c0cad21758d9df53edf9094bdd698c6f74f98bb7e7e5ac275125e",
        "/conf.d": "/home/user/dd-agent",
        "/host/proc": "/proc",
        "/host/sys/fs/cgroup": "/sys/fs/cgroup",
        "/var/run/docker.sock": "/run/docker.sock"
    },
    "VolumesRW": {
        "/checks.d": true,
        "/conf.d": false,
        "/host/proc": false,
        "/host/sys/fs/cgroup": false,
        "/var/run/docker.sock": true
    }
}
]
```

**Comment by ShakataGaNai** [comment]
I'm also running into this issue (Identical error message). My environment is similar, however I'm running this on the host machine rather inside of Docker.

Ansible 2.1.0, Ubuntu 16.04 host, Docker 1.11.2.  Also attempted to provide privileged flag, no change.

**Comment by chouseknecht** [comment]
If I run the following on a Centos 7 vagrant VM using the latest docker_container from source:

```
- name: Start DataDog agent inside container
      docker_container:
        name: dd-agent
        hostname: foo
        image: "datadog/docker-dd-agent:latest"
        restart_policy: always
        restart: yes
        env:
          API_KEY: "asdfadsfs3343raw4t"
        volumes:
          - /var/run/docker.sock:/var/run/docker.sock
          - /proc:/host/proc:ro
          - /sys/fs/cgroup:/host/sys/fs/cgroup:ro
          - "{{ playbook_dir }}/dd-agent:/conf.d:ro"
```

It works. Or, it at least creates the container and puts it in a _running_ state:

```
$ docker ps
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS              PORTS                NAMES
3c223af206fb        datadog/docker-dd-agent:latest   "/entrypoint.sh super"   6 minutes ago       Up 6 minutes        8125/udp, 9001/tcp   dd-agent
```

An inspection of the resulting container reveals the volumes:

```
"Mounts": [
            {
                "Source": "/home/vagrant/dd-agent",
                "Destination": "/conf.d",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Name": "cf66c8260b2a7f3730dbfe7146cf934f0fbc025b3ff4a833ed1fa17058cd1d0a",
                "Source": "/var/lib/docker/volumes/cf66c8260b2a7f3730dbfe7146cf934f0fbc025b3ff4a833ed1fa17058cd1d0a/_data",
                "Destination": "/checks.d",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            },
            {
                "Source": "/var/run/docker.sock",
                "Destination": "/var/run/docker.sock",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            },
            {
                "Source": "/sys/fs/cgroup",
                "Destination": "/host/sys/fs/cgroup",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            },
            {
                "Source": "/proc",
                "Destination": "/host/proc",
                "Mode": "ro",
                "RW": false,
                "Propagation": "rprivate"
            }
        ],
```

**Comment by chouseknecht** [comment]
Ugh. ansibot. 

I don't think this is an issue with docker_container, as I've shown that the volumes in question can actually be mounted when running docker_container inside a traditional VM. How things work in the context of williamyeh/ansible:debian8, I have no idea. 

Also, as I stated, I am running docker_container from source. Lots of bugs have been closed since docker_container landed in 2.1, so it may be that you need to try running with the latest version of docker_container. 

Closing.

**Comment by ShakataGaNai** [comment]
So I wanted to put in my 2 cents again that I had the exact same issue on Ansible 2.1.0 with an Ubuntu 16.04 host. No VM's.

**Comment by chouseknecht** [comment]
@ShakataGaNai I can confirm that it is indeed broken in 2.1.0. If I `pip install ansible==2.1.0` and run the example playbook, I get the reported error.  However, I can also confirm that it is fixed in 2.1.1.0 

To get the fix you must either `pip install ansible==2.1.1.0` or [run from source](http://docs.ansible.com/ansible/intro_installation.html#running-from-source).

If there are other volume related issues, please open a new issue with further details and examples. But I think the specific issue raised here has been addressed.

**Comment by czerasz** [comment]
I can confirm that this issue is not present on `2.1.1.0`. I just pulled the `williamyeh/ansible:debian8`.
